snuba/stateful_consumer/control_protocol.py

```diff
@@ -0,0 +1,172 @@
+
+import jsonschema
+
+from dataclasses import dataclass
+from typing import Any, Mapping, Sequence
+from abc import ABC
+
+from snuba.snapshots import SnapshotId
+from snuba.snapshots.postgres_snapshot import Xid
+
+# TODO: Putting all messages in a single shema makes it hard to interpret errors
+# since the parser will always try all the options and it will rarely tell us which
+# is the actual error because it does not know which is the expected message.
+# We should try to split them into three schemas.
+CONTROL_MSG_SCHEMA = {
+    'anyOf': [
+        {"$ref": "#/definitions/snapshot-init"},
+        {"$ref": "#/definitions/snapshot-abort"},
+        {"$ref": "#/definitions/snapshot-loaded"},
+    ],
+    "definitions": {
+        "base": {
+            "type": "object",
+            "properties": {
+                "snapshot-id": {
+                    "type": "string"
+                }
+            },
+            "required": ["event", "snapshot-id"],
+        },
+        "snapshot-init": {
+            "allOf": [
+                {"$ref": "#/definitions/base"},
+                {
+                    "properties": {
+                        "event": {"const": "snapshot-init"},
+                        "product": {"type": "string"},
+                    },
+                    "required": ["event", "product"],
+                }
+            ]
+        },
+        "snapshot-abort": {
+            "allOf": [
+                {"$ref": "#/definitions/base"},
+                {
+                    "properties": {
+                        "event": {"const": "snapshot-abort"},
+                    },
+                    "required": ["event"],
+                }
+            ]
+        },
+        "dataset": {
+            "type": "object",
+            "properties": {
+                "temp_table": {"type": "string"},
+            },
+            "required": ["temp_table"],
+        },
+        "snapshot-loaded": {
+            "allOf": [
+                {"$ref": "#/definitions/base"},
+                {
+                    "properties": {
+                        "event": {"const": "snapshot-loaded"},
+                        "datasets": {
+                            "type": "object",
+                            "additionalProperties": {
+                                '$ref': '#/definitions/dataset',
+                            },
+                        },
+                        "transaction-info": {
+                            "type": "object",
+                            "properties": {
+                                "xmin": {"type": "number"},
+                                "xmax": {"type": "number"},
+                                "xip-list": {
+                                    "type": "array",
+                                    "items": [
+                                        {"type": "number"}
+                                    ],
+                                },
+                            }
```

Comment:
Shouldn't all of `xmin`/`xmax`/`xip-list` be required?

Reply:
just forgot